tika-parsers/src/main/java/org/apache/tika/parser/captioning/tf/TensorflowRESTCaptioner.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.ByteArrayEntity;
+import org.apache.http.impl.client.DefaultHttpClient;
+import org.apache.tika.config.Field;
+import org.apache.tika.config.Param;
+import org.apache.tika.exception.TikaConfigException;
+import org.apache.tika.exception.TikaException;
+import org.apache.tika.io.IOUtils;
+import org.apache.tika.metadata.Metadata;
+import org.apache.tika.mime.MediaType;
+import org.apache.tika.parser.ParseContext;
+import org.apache.tika.parser.recognition.ObjectRecogniser;
+import org.apache.tika.parser.recognition.RecognisedObject;
+import org.apache.tika.parser.captioning.CaptionObject;
+import org.apache.uima.tools.util.gui.Caption;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.ContentHandler;
+import org.xml.sax.SAXException;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Arrays;
+
+/**
+ * Tensorflow image captioner.
+ * This implementation uses Tensorflow via REST API.
+ * <p>
+ * NOTE : //TODO: link to wiki page here
+ *
+ * @since Apache Tika 1.16
+ */
+public class TensorflowRESTCaptioner implements ObjectRecogniser {
+    private static final Logger LOG = LoggerFactory.getLogger(TensorflowRESTCaptioner.class);
+
+    private static final Set<MediaType> SUPPORTED_MIMES = Collections.unmodifiableSet(
+            new HashSet<>(Arrays.asList(new MediaType[]{
+                    MediaType.image("png"), MediaType.image("jpeg")
+            })));
+
+    private static final String LABEL_LANG = "en";
+
+    @Field
+    private int captions;
+
+    @Field
+    private int maxCaptionLength;
+
+    @Field
+    private URI apiUri;
+
+    @Field
+    private URI healthUri = URI.create("http://localhost:8764/inception/v3/ping");
+
+    private boolean available;
+
+    protected URI getApiUri(Metadata metadata) {
+        return apiUri;
+    }
+
+    @Override
+    public Set<MediaType> getSupportedMimes() {
+        return SUPPORTED_MIMES;
+    }
+
+    @Override
+    public boolean isAvailable() {
+        return available;
+    }
+
+    @Override
+    public void initialize(Map<String, Param> params) throws TikaConfigException {
+        try {
+            DefaultHttpClient client = new DefaultHttpClient();
+            HttpResponse response = client.execute(new HttpGet(healthUri));
+            available = response.getStatusLine().getStatusCode() == 200;
+            LOG.info("Available = {}, API Status = {}", available, response.getStatusLine());
+            apiUri = URI.create(String.format(
+                    "http://localhost:8764/inception/v3/captions?beam_size=%1$d&max_caption_length=%2$d",
```

Comment:
Suggestion:
Instead of having two parameters as `apiUri` and `healthUri`,
let's have a single parameter as `apiBaseUri`, it could take a value like `http://localhost:8764/inception/v3`

then internally we can construct the two other URIs:
+ `healthUri = apiBaseUri + "/ping";`
+ `apiUri = apiBaseUri + String.format("/captions?args=%s", args);`

Here the assumption of fixed relative paths `/ping` and `/captions` seems a reasonable one! So no need to let user modify these two

Thoughts?